ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt

```diff
@@ -0,0 +1,185 @@
+ *           <li><code>local</code>: The host and port of the request uri (e.g. 'www.ktor.io')</li>
+ *           <li><code>method</code>: The http method (e.g. 'GET')</li>
+ *           <li><code>route</code>: The use ktor route used for this request. (e.g. '/some/path/{someParameter}')
+ *           <li><code>status</code>: The http status code or the Exception name thrown during the request (e.g. '200', 'java.lang.IllegalArgumentException')
+ *        <ul>
+ *     <li>
+ *  <ul>
+ */
+class MicrometerMetrics(
+    private val registry: MeterRegistry,
+    timerDistributionConfig: DistributionStatisticConfig,
+    private val tagCustomizer: MicrometerMetrics.TagBuilder.(ApplicationCall) -> Unit
+) {
+
+    private val active = registry.gauge("$baseName.requests.active", AtomicInteger(0))
+
+    init {
+        enableTimerDistributionConfig(timerDistributionConfig)
+    }
+
+    private fun enableTimerDistributionConfig(timerDistributionConfig: DistributionStatisticConfig) {
+        registry.config().meterFilter(object : MeterFilter {
+            override fun configure(id: Meter.Id, config: DistributionStatisticConfig): DistributionStatisticConfig =
+                if (id.name == requestTimerName)
+                    timerDistributionConfig.merge(config)
+                else
+                    config
+        })
+    }
+
+    /**
+     * Configures this Feature
+     * @property registry The meter registry where the meters are registered. Mandatory
+     * @property meterBinders The binders that are automatically bound to the registry. Default: [ClassLoaderMetrics],
+     * [JvmMemoryMetrics], [ProcessorMetrics], [JvmGcMetrics], [ProcessorMetrics], [JvmThreadMetrics], [FileDescriptorMetrics]
+     * @property distributionStatisticConfig configures the histogram and/or percentiles for all request timers. By
+     * default 50%, 90% , 95% and 99% percentiles are configured
+     * @property tagBuilder is called to customize the tags of the timer for each call. by default this is a no-op
+     */
+    class Configuration {
+
+        lateinit var registry: MeterRegistry
+
+        var meterBinders: List<MeterBinder> = listOf(
+            ClassLoaderMetrics(),
+            JvmMemoryMetrics(),
+            JvmGcMetrics(),
+            ProcessorMetrics(),
+            JvmThreadMetrics(),
+            FileDescriptorMetrics()
+        )
+
+        var distributionStatisticConfig: DistributionStatisticConfig =
+            DistributionStatisticConfig.Builder()
+                .percentiles(0.5, 0.9, 0.95, 0.99)
+                .build()
+
+        var tagCustomizer: TagBuilder.(ApplicationCall) -> Unit = {}
+    }
+
+    private fun ApplicationCall.getLocalAddress() = request.local.let { "${it.host}:${it.port}" }
+
+    private fun ApplicationCall.getStatusCode() = response.status()?.value?.toString() ?: "Unknown"
+
+    private fun ApplicationCall.tags(throwable: Throwable? = null) = listOf<Tag>(
+        Tag.of("local", getLocalAddress()),
```

Comment:
The main reason was for orchestration frameworks such as Docker Swarm and Kubernetes, to differentiate between metrics of each replica of a service, but I'd have to look into it a bit more, since it could be that they add these labels themselves...

Reply:
👍 for the @cy6erGn0m suggestion. The user should have the choice of both options. In the case of kubernetes, I think the local address is automatically added as the `instance` tag, so we possibly don't need this tag at all.

Reply:
Rethinking this: The most important information is what the browser requested in the URL. In the example above this would be `my-web-site.com:443`. The physical address of instance where the current server is running, should be discovered by the service discovery or something else. In my currents project its not the physical address but the instance's id (0,1,2,...) which is relevant. 
When the developer needs this information, they can add this information (or whatever they need) and add a MetricFilter that adds these informations as static tags to all metrics of this instance.

Reply:
Then it should be the remote one. Or it should be `localAddress` and `remoteAddress` so user can decide. However this decision is not that easy so a regular user may choose wrong so I'd prefer to have one default since the second could be easily achieved by specifying `tag(name, value)`.

Reply:
I'd suggest not trying to get too smart with adding default common tags. They can vary substantially from runtime to runtime. The user can add common tags via `MeterRegistry#commonTags(...)` at any time.